### ISSUE TITLE
charts appear blur on canvas - make crisp

### ISSUE THREAD

**Issue by anoosurf** [apache/incubator-echarts#13504]
### Version
4.9.0

### Steps to reproduce
show any bar chart on different screen resolutions / pixel ratio it shows blur on edges

### What is expected?
it should be crisp like highcharts.com or svg borders

### What is actually happening?
the dimensions are in float number and not rounded

**Comment by anoosurf** [comment]
![image](https://user-images.githubusercontent.com/4696146/97448896-22b3f400-1957-11eb-9ebb-6c0060c49d09.png)

**Comment by anoosurf** [comment]
![image](https://user-images.githubusercontent.com/4696146/97448984-38291e00-1957-11eb-9850-dd92fc15f6f8.png)

**Comment by anoosurf** [comment]
![image](https://user-images.githubusercontent.com/4696146/97450001-43307e00-1958-11eb-9186-125489a5be92.png)

**Comment by anoosurf** [comment]
maybe a solution could be rounding of the drawing rectangle dimensions and/or shifting of canvas context to adjust for half pixel?
please do assist ASAP as this will make the charts even more great to look at :) 
thanks

**Comment by plainheart** [comment]
Could you please try to check if this issue exists in 5.0? You may use the version 5.0.0-beta.2.

**Comment by anoosurf** [comment]
yes i just checked right now .. the issue exists in the latest beat
versionof 5.0.0 beta-2 also


i tried a quick fix like this below to make the rectangle sharper  .. but
it doesn't work everytime :(



   rect: function (x, y, w, h) {
            //edit 22/10/2020
            // for columns
            x = Math.round(x) + 0.5;
            y = Math.round(y) + 0.5;
            w = Math.round(w);
            h = Math.round(h);

            //if (x % 2 !== 0) { // hack  for odd heights
            //    x = Math.round(x) ;
            //}
            //if (y % 2 !== 0) { // hack  for odd heights
            //    y = Math.round(y) ;
            //}

            //if (h % 2 !== 0) { // hack  for odd heights
            //    x = Math.round(x);
            //    y = Math.round(y);
            //}
            if (h >= 15) { // hack  for bars
                x = Math.round(x);
                y = Math.round(y);
            }

            //console.log('x:' + x + ' y:' + y + ' w:' + w + ' h:' + h);


            this._ctx && this._ctx.rect(x, y, w, h);
            this.addData(CMD.R, x, y, w, h);
            return this;
        },

**Comment by plainheart** [comment]
Thanks for the check. Just found you had gotten involved in this issue in #13504. It may still take some time to fix it.

**Comment by TaylorPzreal** [comment]
Is there has a version that doesn't has this error?
my echarts version is 4.9.0 and has this same error;

**Comment by TaylorPzreal** [comment]
After i read ```zrender``` source code, i found how to trigger this problem in my case;

First the ```zrender``` [src/config.ts](https://github.com/ecomfe/zrender/blob/master/src/config.ts) portion code:

```ts
let dpr = 1;

// If in browser environment
if (typeof window !== 'undefined') {
    dpr = Math.max(window.devicePixelRatio 
    	|| ((window.screen as any).deviceXDPI / (window.screen as any).logicalXDPI) 
    	|| 1, 1);
}
```

As you can see, ```zrender``` use ```window.devicePixelRatio ``` to set pixel almost any time in today browser if you don't config this, and my case is:

First i displayed my webpage on a low pixel screen(1PX), and then i dragged the browser to mac screen(2PX), so i found this problem.

So it's okay for me. Just for sharing my case.

**Comment by rajsite** [comment]
I see similar issues in the [charts examples](https://echarts.apache.org/examples/en/editor.html?c=area-stack-gradient) if I zoom the browser:

![echarts](https://user-images.githubusercontent.com/1588923/187729142-390e632a-c567-448a-85cc-2851df13546c.gif)

Notice that when I refresh the page the echart reloads and sees the new devicePixelRatio and looks sharp again. I think the issue is that echarts needs a way to allow for monitoring changes to the devicePixelRatio dynamically.

MDN has an example for [Monitoring screen resolution or zoom level changes](https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio#monitoring_screen_resolution_or_zoom_level_changes) which should help for the browser zoom use-case I show.

**Comment by mballav** [comment]
I am facing the same issue when trying to produce high-quality PDFs using puppeteer. While other items on the web page look crisp even when zoomed in to 300% or more, the charts look blurry, as you can see below:

![image](https://user-images.githubusercontent.com/18305518/217633685-e47cc878-bee2-422e-9b5f-19e3376c0e64.png)

Could you please take this enhancement up in the next release? It's really important for us to deliver high-quality reports to our clients.

**Comment by KamilAndrzejczukInspeerity** [comment]
Any update on that? I face same issue on v5.5.1

**Comment by syedwj** [comment]
![image](https://github.com/user-attachments/assets/ba4c4432-9744-49cb-9bc8-83b411bd514c)
image are blury any one who find out solution
